TildeSql/Internal/QueryEngine.cs

```diff
@@ -203,51 +199,81 @@ public async IAsyncEnumerable<T> GetResult<T>(IQuery query)
-                }
-                else {
-                    this.queryForwardMap[original] = [executed, remaining];
-                }
+                this.queryForwardMap[original] = [executed, remaining];
             }
 
             queriesStillToExecute = identityMapExecutionResult.NonExecutedQueries.Union(identityMapExecutionResult.PartiallyExecutedQueries.Select(pq => pq.Remaining)).ToList();
 
-            if (queriesStillToExecute.Any()) {
-                foreach (var entry in this.cacheExecutors.AsSmartEnumerable()) {
-                    var cacheExecutionResult = await entry.Value.ExecuteAsync(queriesStillToExecute, cancellationToken);
-                    foreach (var executedQuery in cacheExecutionResult.ExecutedQueries) {
-                        this.cacheExecutorQueries[entry.Index].Add(executedQuery);
-                    }
+            if (queriesStillToExecute.Any() && this.cacheExecutor != null) {
+                await ExecuteAgainstCacheAsync().ToArrayAsync(cancellationToken);
+            }
 
-                    foreach (var (original, executed, remaining) in cacheExecutionResult.PartiallyExecutedQueries) {
-                        if (this.queryForwardMap.ContainsKey(original)) {
-                            this.queryForwardMap.Add(original, [executed, remaining]);
-                        }
-                        else {
-                            this.queryForwardMap[original] = [executed, remaining];
-                        }
-                    }
+            async IAsyncEnumerable<IQuery> ExecuteAgainstCacheAsync() {
+                var cacheExecutionResult = await this.cacheExecutor.ExecuteAsync(queriesStillToExecute, cancellationToken);
+                foreach (var executedQuery in cacheExecutionResult.ExecutedQueries) {
+                    this.cacheExecutorQueries.Add(executedQuery);
+                    yield return executedQuery;
+                }
 
-                    queriesStillToExecute = cacheExecutionResult.NonExecutedQueries.Union(cacheExecutionResult.PartiallyExecutedQueries.Select(pq => pq.Remaining)).ToList();
-                    if (!queriesStillToExecute.Any()) {
-                        break;
-                    }
+                foreach (var (original, executed, remaining) in cacheExecutionResult.PartiallyExecutedQueries) {
+                    this.queryForwardMap[original] = [executed, remaining];
                 }
+
+                queriesStillToExecute = cacheExecutionResult.NonExecutedQueries.Union(cacheExecutionResult.PartiallyExecutedQueries.Select(pq => pq.Remaining)).ToList();
             }
 
             if (queriesStillToExecute.Any()) {
                 if (this.persistenceQueryExecutor == null) {
                     throw new Exception("No persistence query mechanism has been configured");
                 }
 
+                if (this.cacheExecutor != null) {
+                    // stampede protection, we take a lock on all cacheable queries
```

Comment:
This is cool but do we need it?

The mutex will only allow one execution path through ExecuteAsync at a time. I think it's not needed at all because we only have one thing executing at a time.

If we do have it, do we need some deadlock protection?
Consider A and B is queried simultaneously as B and A. A is locked to 1, B is locked to 2 neither can get the lock for the second query.

Reply:
We spoke - this is locking across the process as opposed to across the session